[Navigation] Initial implementations of traversal methods

TingPing opened the pull request:
#### a0dd312f71c532a17e69e5703543708c7af68ba9
<pre>
[Navigation] Initial implementations of traversal methods
<a href="https://bugs.webkit.org/show_bug.cgi?id=268465">https://bugs.webkit.org/show_bug.cgi?id=268465</a>

Reviewed by Alex Christensen.

This is the very beginnings of the traversal methods.

This implements a lot of the error handling but does not
fully implement traversal, only doing basic manual
traversals which can be useful for testing.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigate-from-initial-about-blank-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-cross-origin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/navigate-from-initial-about-blank-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/navigate-from-initial-about-blank-gc-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/navigate-from-initial-about-blank-src-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/reload-navigation-timing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/transition-cross-document-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/per-entry-events/dispose-same-document-reload-with-intercept-expected.txt:
* Source/WebCore/page/Navigation.cpp:
(WebCore::createEarlyErrorResult):
(WebCore::Navigation::reload):
(WebCore::Navigation::navigate):
(WebCore::Navigation::performTraversal):
(WebCore::Navigation::findEntryByKey):
(WebCore::Navigation::traverseTo):
(WebCore::Navigation::back):
(WebCore::Navigation::forward):
* Source/WebCore/page/Navigation.h:
* Source/WebCore/page/Navigation.idl:

Canonical link: <a href="https://commits.webkit.org/275054@main">https://commits.webkit.org/275054@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1b696706e67e4300c47e815992de67a382115c9e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39630 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18609 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41986 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42165 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35530 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41936 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21535 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15938 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33085 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40204 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15722 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34295 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13612 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13620 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35249 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43442 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36012 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35582 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39370 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14442 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11894 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37636 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16048 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9137 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16097 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/15705 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rwlbuis commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -75,19 +83,32 @@ EventTargetInterface Navigation::eventTargetInterface() const
 {
-    // FIXME: keep track of promises to resolve later.
-    Ref entry = NavigationHistoryEntry::create(scriptExecutionContext(), { });
     auto globalObject = committed->globalObject();
```
We should probably assert that committed and finished globalObjects are equal.

rwlbuis commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -75,19 +83,32 @@ EventTargetInterface Navigation::eventTargetInterface() const
 
-Navigation::Result Navigation::navigate(const String& /* url */, NavigateOptions&&, Ref<DeferredPromise>&& committed, Ref<DeferredPromise>&& finished)
+// https://html.spec.whatwg.org/multipage/nav-history-apis.html#navigation-api-early-error-result
+static Navigation::Result createEarlyErrorResult(Ref<DeferredPromise> committed, Ref<DeferredPromise> finished, Exception&& exception)
```
I think the Early in the name is kind of confusing, an error may be generated relatively late. Probably best to just remove the 'Early' reference.

rwlbuis commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -97,37 +118,109 @@ Navigation::Result Navigation::reload(ReloadOptions&&, Ref<DeferredPromise>&& co
+Navigation::Result Navigation::navigate(ScriptExecutionContext& scriptExecutionContext, const String& url, NavigateOptions&& options, Ref<DeferredPromise>&& committed, Ref<DeferredPromise>&& finished)
 {
+    auto currentURL = scriptExecutionContext.url();
+    auto fullURL = URL { currentURL, url };
```
Should it be something like newURL, in contrast to currentURL?

rwlbuis commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -97,37 +118,109 @@ Navigation::Result Navigation::reload(ReloadOptions&&, Ref<DeferredPromise>&& co
-    finished->resolve<IDLInterface<NavigationHistoryEntry>>(entry.get());
-    return result;
+    RELEASE_ASSERT(m_entries.size() <= INT_MAX);
+    if (m_currentEntryIndex < 0 || m_currentEntryIndex == static_cast<int>(m_entries.size()))
```
I think the comparison may be wrong, for 1 entry we'd get 0 == 1 and not throw an exception, yet we still should not be allowed to go forward.
I had the same problem in https://github.com/WebKit/WebKit/pull/23578, you can make m_currentEntryIndex std::optional as a fix. For one thing the spec seems to have 1-based arrays, and then the above check would just work, but with 0-based we have to do size -1, think about unsigned vs signed etc. Probably best to avoid such problems.

TingPing committed:
[Navigation] Initial implementations of traversal methods
https://bugs.webkit.org/show_bug.cgi?id=268465

Reviewed by Alex Christensen.

This is the very beginnings of the traversal methods.

This implements a lot of the error handling but does not
fully implement traversal, only doing basic manual
traversals which can be useful for testing.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigate-from-initial-about-blank-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-cross-origin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/navigate-from-initial-about-blank-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/navigate-from-initial-about-blank-gc-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/navigate-from-initial-about-blank-src-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/reload-navigation-timing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/transition-cross-document-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/per-entry-events/dispose-same-document-reload-with-intercept-expected.txt:
* Source/WebCore/page/Navigation.cpp:
(WebCore::createEarlyErrorResult):
(WebCore::Navigation::reload):
(WebCore::Navigation::navigate):
(WebCore::Navigation::performTraversal):
(WebCore::Navigation::findEntryByKey):
(WebCore::Navigation::traverseTo):
(WebCore::Navigation::back):
(WebCore::Navigation::forward):
* Source/WebCore/page/Navigation.h:
* Source/WebCore/page/Navigation.idl:

Canonical link: https://commits.webkit.org/275054@main
